File: php/sync.php
#!/usr/bin/env php
<?php
/**
 * ⚠️ WARNUNG: SICHERHEITSRISIKO ⚠️
 * 
 * Dieses Script verarbeitet SENSIBLE PASSWÖRTER auf einem SERVER!
 * 
 * GEFAHREN:
 * - Bei Server-Kompromittierung = alle Passwörter weg
 * - PHP-Fehler könnten sensible Daten loggen
 * - Zugriff über Web = hohes Risiko
 * - Passwörter werden im RAM verarbeitet
 * 
 * ANFORDERUNGEN:
 * - ✅ Nur auf VERTRAUENSWÜRDIGEM Server verwenden (VPS/Dedicated)
 * - ✅ .htaccess mit Zugriffsbeschränkung (Apache)
 * - ✅ Nginx: location block mit deny all
 * - ✅ HTTPS zwingend erforderlich
 * - ✅ Regelmäßige Sicherheits-Updates
 * - ✅ Logs regelmäßig prüfen
 * - ✅ KeePassXC-CLI muss auf Server installiert sein
 * - ✅ Script sollte NUR über Cronjob laufen (nicht über Web)
 * 
 * EMPFOHLEN: Für Desktop-Nutzung besser Python-Variante verwenden (sicherer)
 * 
 * @package KeePassSync
 * @version 1.0.0
 */

// Nur für CLI-Ausführung erlauben (nicht über Web)
if (php_sapi_name() !== 'cli') {
    http_response_code(403);
    die("ERROR: This script must be run from command line (CLI) only!\n" .
        "For security reasons, web access is forbidden.\n");
}

// Fehlerbehandlung
error_reporting(E_ALL);
ini_set('display_errors', 0); // Keine Fehler in Output (nur Logs)
ini_set('log_errors', 1);
ini_set('error_log', __DIR__ . '/php_error.log');

// Pfade
$SCRIPT_DIR = __DIR__;
$BASE_DIR = dirname($SCRIPT_DIR);
$CONFIG_FILE = $BASE_DIR . '/config.json';
$LOG_FILE = $BASE_DIR . '/sync_log.txt';
$FTP_LOG = $BASE_DIR . '/ftp_log.txt';

// In Basis-Verzeichnis wechseln
chdir($BASE_DIR);

/**
 * Schreibe Nachricht ins Log
 */
function write_log($message, $log_file = null) {
    global $LOG_FILE;
    if ($log_file === null) {
        $log_file = $LOG_FILE;
    }
    
    $timestamp = date('Y-m-d H:i:s');
    $log_message = "$timestamp $message\n";
    echo $log_message;
    
    @file_put_contents($log_file, $log_message, FILE_APPEND);
}

/**
 * Lade Konfiguration
 */
function load_config() {
    global $CONFIG_FILE;
    
    if (!file_exists($CONFIG_FILE)) {
        write_log("ERROR: Configuration file not found: $CONFIG_FILE");
        exit(1);
    }
    
    $config = @json_decode(file_get_contents($CONFIG_FILE), true);
    if ($config === null) {
        write_log("ERROR: Failed to parse configuration file");
        exit(1);
    }
    
    return $config;
}

/**
 * Finde KeePassXC-CLI
 */
function find_keepassxc($hint = 'keepassxc-cli') {
    // Versuche direkt
    if ($hint && is_executable($hint)) {
        return $hint;
    }
    
    // Suche im PATH
    $paths = explode(':', getenv('PATH') ?: '');
    foreach ($paths as $path) {
        $full_path = $path . '/' . $hint;
        if (is_executable($full_path)) {
            return $full_path;
        }
    }
    
    // Versuche which/whereis
    $which = trim(shell_exec("which $hint 2>/dev/null"));
    if ($which && is_executable($which)) {
        return $which;
    }
    
    return null;
}

/**
 * Erstelle Backup
 */
function create_backup($local_db, $backup_dir) {
    if (!file_exists($local_db)) {
        write_log("WARNING: Local database not found: $local_db");
        return false;
    }
    
    if (!is_dir($backup_dir)) {
        @mkdir($backup_dir, 0755, true);
    }
    
    $backup_file = $backup_dir . '/' . 'keepass_passwords_' . date('Ymd') . '.kdbx';
    
    if (@copy($local_db, $backup_file)) {
        write_log("Backup successfully created: " . basename($backup_file));
        return true;
    } else {
        write_log("WARNING: Could not create backup");
        return false;
    }
}

/**
 * Räume alte Backups auf
 */
function cleanup_backups($backup_dir, $max_backups = 2) {
    $backups = glob($backup_dir . '/keepass_passwords_*.kdbx');
    if (count($backups) <= $max_backups) {
        return;
    }
    
    // Sortiere nach Änderungszeit
    usort($backups, function($a, $b) {
        return filemtime($b) - filemtime($a);
    });
    
    // Lösche alte Backups
    foreach (array_slice($backups, $max_backups) as $old_backup) {
        @unlink($old_backup);
        write_log("Old backup deleted: " . basename($old_backup));
    }
}

/**
 * Download von FTP-Server
 */
function download_ftp($host, $user, $password, $remote_path, $local_file) {
    write_log("Starting download from server...");
    
    $conn = @ftp_connect($host);
    if (!$conn) {
        write_log("ERROR: Could not connect to FTP server: $host");
        return false;
    }
    
    $login = @ftp_login($conn, $user, $password);
    if (!$login) {
        write_log("ERROR: FTP login failed");
        @ftp_close($conn);
        return false;
    }
    
    // Aktiviere passiven Modus
    @ftp_pasv($conn, true);
    
    $result = @ftp_get($conn, $local_file, $remote_path, FTP_BINARY);
    @ftp_close($conn);
    
    if ($result && file_exists($local_file)) {
        write_log("Download successful");
        return true;
    } else {
        write_log("ERROR: Download failed");
        return false;
    }
}

/**
 * Upload zu FTP-Server
 */
function upload_ftp($host, $user, $password, $remote_path, $local_file) {
    write_log("Starting upload to server...");
    
    $conn = @ftp_connect($host);
    if (!$conn) {
        write_log("ERROR: Could not connect to FTP server: $host");
        return false;
    }
    
    $login = @ftp_login($conn, $user, $password);
    if (!$login) {
        write_log("ERROR: FTP login failed");
        @ftp_close($conn);
        return false;
    }
    
    // Aktiviere passiven Modus
    @ftp_pasv($conn, true);
    
    $result = @ftp_put($conn, $remote_path, $local_file, FTP_BINARY);
    @ftp_close($conn);
    
    if ($result) {
        write_log("Upload successful");
        return true;
    } else {
        write_log("ERROR: Upload failed");
        return false;
    }
}

/**
 * Merge Datenbanken mit KeePassXC-CLI
 */
function merge_databases($keepassxc, $local_db, $temp_db, $password) {
    write_log("Performing merge...");
    
    if (!file_exists($temp_db)) {
        write_log("ERROR: Temporary file not found: $temp_db");
        return false;
    }
    
    $cmd = escapeshellcmd($keepassxc) . 
           ' merge -s ' . 
           escapeshellarg($local_db) . ' ' . 
           escapeshellarg($temp_db) . 
           ' --same-credentials 2>&1';
    
    $descriptorspec = [
        0 => ['pipe', 'r'],  // stdin
        1 => ['pipe', 'w'],  // stdout
        2 => ['pipe', 'w']   // stderr
    ];
    
    $process = @proc_open($cmd, $descriptorspec, $pipes);
    if (!is_resource($process)) {
        write_log("ERROR: Could not start KeePassXC-CLI");
        return false;
    }
    
    // Passwort übergeben
    fwrite($pipes[0], $password);
    fclose($pipes[0]);
    
    $output = stream_get_contents($pipes[1]);
    $error = stream_get_contents($pipes[2]);
    fclose($pipes[1]);
    fclose($pipes[2]);
    
    $return_code = proc_close($process);
    
    if ($return_code === 0) {
        write_log("Merge completed successfully. Local file updated.");
        return true;
    } else {
        write_log("ERROR: Merge failed: $error");
        return false;
    }
}

/**
 * Hauptfunktion
 */
function main() {
    global $CONFIG_FILE, $BASE_DIR;
    
    write_log("=== KeePass Sync - PHP CLI ===");
    
    // Warnung ausgeben
    write_log("WARNING: This script processes sensitive passwords on a server!");
    write_log("Ensure proper security measures are in place.");
    
    // Konfiguration laden
    $config = load_config();
    
    // KeePassXC-CLI finden
    $keepassxc_path = $config['keepass']['keepassXCPath'] ?? 'keepassxc-cli';
    $keepassxc = find_keepassxc($keepassxc_path);
    
    if (!$keepassxc) {
        write_log("ERROR: KeePassXC-CLI not found. Install: apt install keepassxc");
        write_log("Or set correct path in config.json");
        exit(1);
    }
    
    write_log("KeePassXC-CLI found: $keepassxc");
    
    // Pfade
    $local_db = $config['local']['localPath'];
    $temp_db = $config['local']['tempPath'];
    $backup_dir = $config['local']['backupDir'];
    $max_backups = $config['local']['maxBackups'] ?? 2;
    $db_password = $config['keepass']['databasePassword'];
    
    // FTP-Konfiguration
    $ftp_config = $config['ftp'];
    $host = $ftp_config['host'];
    $user = $ftp_config['user'];
    $password = $ftp_config['password'];
    $remote_path = $ftp_config['remotePath'];
    $protocol = strtolower($ftp_config['type'] ?? 'ftp');
    
    // Nur FTP wird unterstützt (SFTP braucht ssh2 Extension)
    if ($protocol !== 'ftp') {
        write_log("WARNING: PHP-Variante unterstützt nur FTP. Für SFTP/SMB/SCP nutze Python-Variante.");
        write_log("Protocol '$protocol' wird als FTP behandelt.");
    }
    
    // Backup erstellen
    write_log("Creating backup...");
    create_backup($local_db, $backup_dir);
    cleanup_backups($backup_dir, $max_backups);
    
    // Download
    if (!download_ftp($host, $user, $password, $remote_path, $temp_db)) {
        exit(1);
    }
    
    // Merge
    if (!merge_databases($keepassxc, $local_db, $temp_db, $db_password)) {
        exit(1);
    }
    
    // Upload
    if (!upload_ftp($host, $user, $password, $remote_path, $local_db)) {
        exit(1);
    }
    
    // Aufräumen
    if (file_exists($temp_db)) {
        @unlink($temp_db);
    }
    
    write_log("Synchronization completed.");
}

// Ausführen
try {
    main();
} catch (Exception $e) {
    write_log("FATAL ERROR: " . $e->getMessage());
    exit(1);
}

File: go/sync.go
package main

/*
KeePass Sync - Go (Golang) Version
Cross-Platform: Linux, Windows, macOS

Compile:
  go build -o keepass-sync sync.go

Or for specific platform:
  GOOS=windows GOARCH=amd64 go build -o keepass-sync.exe sync.go
  GOOS=linux GOARCH=amd64 go build -o keepass-sync sync.go
  GOOS=darwin GOARCH=amd64 go build -o keepass-sync sync.go

Features:
  - Single binary (no dependencies needed)
  - Fast execution
  - Cross-platform
  - CLI arguments (--test, --status, --watch, etc.)
  - Retry logic with exponential backoff
  - Supports FTP, SFTP, SMB, SCP protocols
*/

import (
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

// Config structure
type Config struct {
	FTP struct {
		Host       string `json:"host"`
		User       string `json:"user"`
		Password   string `json:"password"`
		Type       string `json:"type"`
		RemotePath string `json:"remotePath"`
		Port       int    `json:"port"`
		Share      string `json:"share"`
		Domain     string `json:"domain"`
	} `json:"ftp"`
	Local struct {
		LocalPath  string `json:"localPath"`
		TempPath   string `json:"tempPath"`
		BackupDir  string `json:"backupDir"`
		MaxBackups int    `json:"maxBackups"`
	} `json:"local"`
	Keepass struct {
		DatabasePassword string `json:"databasePassword"`
		KeepassXCPath    string `json:"keepassXCPath"`
	} `json:"keepass"`
	Settings struct {
		Debug        bool   `json:"debug"`
		Language     string `json:"language"`
		CleanupLogs  bool   `json:"cleanupLogs"`
		MaxLogAge    int    `json:"maxLogAgeDays"`
		MaxRetries   int    `json:"max_retries"`
		RetryDelay   int    `json:"retry_delay"`
		WatchDelay   int    `json:"watch_delay"`
	} `json:"settings"`
}

var (
	configFile = "config.json"
	logFile    = "sync_log.txt"
	ftpLog     = "ftp_log.txt"
	config     Config
)

func writeLog(message string) {
	timestamp := time.Now().Format("2006-01-02 15:04:05")
	logMessage := fmt.Sprintf("%s %s", timestamp, message)
	fmt.Println(logMessage)
	
	f, err := os.OpenFile(logFile, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return
	}
	defer f.Close()
	fmt.Fprintln(f, logMessage)
}

func loadConfig() error {
	data, err := os.ReadFile(configFile)
	if err != nil {
		return fmt.Errorf("configuration file not found: %s", configFile)
	}
	
	if err := json.Unmarshal(data, &config); err != nil {
		return fmt.Errorf("failed to parse configuration: %v", err)
	}
	
	// Set defaults
	if config.Settings.MaxRetries == 0 {
		config.Settings.MaxRetries = 3
	}
	if config.Settings.RetryDelay == 0 {
		config.Settings.RetryDelay = 5
	}
	if config.Settings.WatchDelay == 0 {
		config.Settings.WatchDelay = 30
	}
	if config.Local.MaxBackups == 0 {
		config.Local.MaxBackups = 2
	}
	
	return nil
}

func findExecutable(name string) string {
	// Try direct path first
	if _, err := os.Stat(name); err == nil {
		return name
	}
	
	// Search in PATH
	path := os.Getenv("PATH")
	paths := strings.Split(path, ":")
	if os.PathSeparator == ';' {
		paths = strings.Split(path, ";") // Windows
	}
	
	for _, p := range paths {
		fullPath := filepath.Join(p, name)
		if _, err := os.Stat(fullPath); err == nil {
			return fullPath
		}
	}
	
	// Try which/where
	cmd := exec.Command("which", name)
	if output, err := cmd.Output(); err == nil {
		return strings.TrimSpace(string(output))
	}
	
	return ""
}

func createBackup(localDB, backupDir string) bool {
	writeLog("Creating backup...")
	
	if _, err := os.Stat(localDB); os.IsNotExist(err) {
		writeLog("WARNING: Local database not found")
		return false
	}
	
	if err := os.MkdirAll(backupDir, 0755); err != nil {
		writeLog(fmt.Sprintf("WARNING: Could not create backup directory: %v", err))
		return false
	}
	
	backupFile := filepath.Join(backupDir, fmt.Sprintf("keepass_passwords_%s.kdbx", time.Now().Format("20060102")))
	
	if err := copyFile(localDB, backupFile); err != nil {
		writeLog(fmt.Sprintf("WARNING: Could not create backup: %v", err))
		return false
	}
	
	writeLog("Backup successfully created")
	return true
}

func copyFile(src, dst string) error {
	source, err := os.Open(src)
	if err != nil {
		return err
	}
	defer source.Close()
	
	destination, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer destination.Close()
	
	_, err = io.Copy(destination, source)
	return err
}

func cleanupBackups(backupDir string, maxBackups int) {
	files, err := filepath.Glob(filepath.Join(backupDir, "keepass_passwords_*.kdbx"))
	if err != nil {
		return
	}
	
	if len(files) <= maxBackups {
		return
	}
	
	// Sort by modification time
	type fileInfo struct {
		path string
		time time.Time
	}
	
	fileInfos := make([]fileInfo, 0, len(files))
	for _, f := range files {
		info, err := os.Stat(f)
		if err != nil {
			continue
		}
		fileInfos = append(fileInfos, fileInfo{
			path: f,
			time: info.ModTime(),
		})
	}
	
	// Sort descending (newest first)
	for i := 0; i < len(fileInfos)-1; i++ {
		for j := i + 1; j < len(fileInfos); j++ {
			if fileInfos[i].time.Before(fileInfos[j].time) {
				fileInfos[i], fileInfos[j] = fileInfos[j], fileInfos[i]
			}
		}
	}
	
	// Delete old backups
	for i := maxBackups; i < len(fileInfos); i++ {
		if err := os.Remove(fileInfos[i].path); err == nil {
			writeLog(fmt.Sprintf("Old backup deleted: %s", filepath.Base(fileInfos[i].path)))
		}
	}
}

func downloadFile(host, user, password, remotePath, tempFile, protocol string, maxRetries, retryDelay int, port int, share, domain string) bool {
	for attempt := 0; attempt < maxRetries; attempt++ {
		if attempt > 0 {
			delay := retryDelay * (1 << uint(attempt-1))
			if delay > 60 {
				delay = 60
			}
			writeLog(fmt.Sprintf("Retry %d/%d in %d seconds...", attempt, maxRetries-1, delay))
			time.Sleep(time.Duration(delay) * time.Second)
		}
		
		var success bool
		switch protocol {
		case "ftp", "sftp":
			success = downloadFTP(host, user, password, remotePath, tempFile, protocol == "sftp", port)
		case "smb":
			success = downloadSMB(host, share, user, password, remotePath, tempFile, domain)
		case "scp":
			success = downloadSCP(host, user, password, remotePath, tempFile, port)
		default:
			writeLog(fmt.Sprintf("Unknown protocol: %s", protocol))
			return false
		}
		
		if success {
			return true
		}
	}
	
	writeLog(fmt.Sprintf("Download failed after %d attempts", maxRetries))
	return false
}

func downloadFTP(host, user, password, remotePath, tempFile string, sftp bool, port int) bool {
	writeLog("Starting download from server...")
	
	// Use lftp (if available) for FTP/SFTP
	lftpPath := findExecutable("lftp")
	if lftpPath != "" {
		var url string
		if sftp {
			url = fmt.Sprintf("sftp://%s", host)
		} else {
			url = fmt.Sprintf("ftp://%s", host)
		}
		
		remoteDir := filepath.Dir(remotePath)
		remoteFile := filepath.Base(remotePath)
		
		script := []string{}
		if sftp {
			script = append(script, `set sftp:connect-program "ssh -a -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"`)
		}
		if remoteDir != "." && remoteDir != "/" {
			script = append(script, fmt.Sprintf("cd %s", remoteDir))
		}
		script = append(script, fmt.Sprintf("get %s -o %s", remoteFile, tempFile))
		script = append(script, "quit")
		
		cmd := exec.Command(lftpPath, "-u", fmt.Sprintf("%s,%s", user, password), url)
		cmd.Stdin = strings.NewReader(strings.Join(script, "\n") + "\n")
		
		if err := cmd.Run(); err != nil {
			writeLog(fmt.Sprintf("Download failed: %v", err))
			return false
		}
		
		if _, err := os.Stat(tempFile); err == nil {
			writeLog("Download successful")
			return true
		}
	}
	
	writeLog("ERROR: lftp not found. Install: sudo apt install lftp")
	return false
}

func downloadSMB(host, share, user, password, remotePath, tempFile, domain string) bool {
	writeLog("Starting download from server...")
	
	smbclientPath := findExecutable("smbclient")
	if smbclientPath == "" {
		writeLog("ERROR: smbclient not found. Install: sudo apt install samba-common")
		return false
	}
	
	smbURL := fmt.Sprintf("//%s/%s", host, share)
	cmd := exec.Command(smbclientPath, smbURL, "-U", fmt.Sprintf("%s\\%s%%%s", domain, user, password), "-c", fmt.Sprintf(`get "%s" "%s"`, remotePath, tempFile))
	
	if err := cmd.Run(); err != nil {
		writeLog(fmt.Sprintf("Download failed: %v", err))
		return false
	}
	
	if _, err := os.Stat(tempFile); err == nil {
		writeLog("Download successful")
		return true
	}
	
	return false
}

func downloadSCP(host, user, password, remotePath, tempFile string, port int) bool {
	writeLog("Starting download from server...")
	
	sshpassPath := findExecutable("sshpass")
	if sshpassPath == "" {
		writeLog("ERROR: sshpass not found. Install: sudo apt install sshpass")
		return false
	}
	
	scpPath := findExecutable("scp")
	if scpPath == "" {
		writeLog("ERROR: scp not found")
		return false
	}
	
	cmd := exec.Command(sshpassPath, "-p", password, scpPath, "-P", strconv.Itoa(port),
		"-o", "StrictHostKeyChecking=no",
		"-o", "UserKnownHostsFile=/dev/null",
		fmt.Sprintf("%s@%s:%s", user, host, remotePath),
		tempFile)
	
	if err := cmd.Run(); err != nil {
		writeLog(fmt.Sprintf("Download failed: %v", err))
		return false
	}
	
	if _, err := os.Stat(tempFile); err == nil {
		writeLog("Download successful")
		return true
	}
	
	return false
}

func uploadFile(host, user, password, remotePath, localFile, protocol string, maxRetries, retryDelay int, port int, share, domain string) bool {
	for attempt := 0; attempt < maxRetries; attempt++ {
		if attempt > 0 {
			delay := retryDelay * (1 << uint(attempt-1))
			if delay > 60 {
				delay = 60
			}
			writeLog(fmt.Sprintf("Retry %d/%d in %d seconds...", attempt, maxRetries-1, delay))
			time.Sleep(time.Duration(delay) * time.Second)
		}
		
		var success bool
		switch protocol {
		case "ftp", "sftp":
			success = uploadFTP(host, user, password, remotePath, localFile, protocol == "sftp", port)
		case "smb":
			success = uploadSMB(host, share, user, password, remotePath, localFile, domain)
		case "scp":
			success = uploadSCP(host, user, password, remotePath, localFile, port)
		default:
			writeLog(fmt.Sprintf("Unknown protocol: %s", protocol))
			return false
		}
		
		if success {
			return true
		}
	}
	
	writeLog(fmt.Sprintf("Upload failed after %d attempts", maxRetries))
	return false
}

func uploadFTP(host, user, password, remotePath, localFile string, sftp bool, port int) bool {
	writeLog("Starting upload to server...")
	
	lftpPath := findExecutable("lftp")
	if lftpPath != "" {
		var url string
		if sftp {
			url = fmt.Sprintf("sftp://%s", host)
		} else {
			url = fmt.Sprintf("ftp://%s", host)
		}
		
		remoteDir := filepath.Dir(remotePath)
		remoteFile := filepath.Base(remotePath)
		
		script := []string{}
		if sftp {
			script = append(script, `set sftp:connect-program "ssh -a -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null"`)
		}
		if remoteDir != "." && remoteDir != "/" {
			script = append(script, fmt.Sprintf("cd %s", remoteDir))
		}
		script = append(script, fmt.Sprintf("put %s -o %s", localFile, remoteFile))
		script = append(script, "quit")
		
		cmd := exec.Command(lftpPath, "-u", fmt.Sprintf("%s,%s", user, password), url)
		cmd.Stdin = strings.NewReader(strings.Join(script, "\n") + "\n")
		
		if err := cmd.Run(); err != nil {
			writeLog(fmt.Sprintf("Upload failed: %v", err))
			return false
		}
		
		writeLog("Upload successful")
		return true
	}
	
	writeLog("ERROR: lftp not found")
	return false
}

func uploadSMB(host, share, user, password, remotePath, localFile, domain string) bool {
	writeLog("Starting upload to server...")
	
	smbclientPath := findExecutable("smbclient")
	if smbclientPath == "" {
		writeLog("ERROR: smbclient not found")
		return false
	}
	
	smbURL := fmt.Sprintf("//%s/%s", host, share)
	cmd := exec.Command(smbclientPath, smbURL, "-U", fmt.Sprintf("%s\\%s%%%s", domain, user, password), "-c", fmt.Sprintf(`put "%s" "%s"`, localFile, remotePath))
	
	if err := cmd.Run(); err != nil {
		writeLog(fmt.Sprintf("Upload failed: %v", err))
		return false
	}
	
	writeLog("Upload successful")
	return true
}

func uploadSCP(host, user, password, remotePath, localFile string, port int) bool {
	writeLog("Starting upload to server...")
	
	sshpassPath := findExecutable("sshpass")
	scpPath := findExecutable("scp")
	if sshpassPath == "" || scpPath == "" {
		writeLog("ERROR: sshpass or scp not found")
		return false
	}
	
	cmd := exec.Command(sshpassPath, "-p", password, scpPath, "-P", strconv.Itoa(port),
		"-o", "StrictHostKeyChecking=no",
		"-o", "UserKnownHostsFile=/dev/null",
		localFile,
		fmt.Sprintf("%s@%s:%s", user, host, remotePath))
	
	if err := cmd.Run(); err != nil {
		writeLog(fmt.Sprintf("Upload failed: %v", err))
		return false
	}
	
	writeLog("Upload successful")
	return true
}

func mergeDatabases(keepassxc, localDB, tempDB, password string) bool {
	writeLog("Performing merge...")
	
	if _, err := os.Stat(tempDB); os.IsNotExist(err) {
		writeLog("ERROR: Temporary file not found")
		return false
	}
	
	cmd := exec.Command(keepassxc, "merge", "-s", localDB, tempDB, "--same-credentials")
	stdin, err := cmd.StdinPipe()
	if err != nil {
		writeLog(fmt.Sprintf("ERROR: Could not start merge: %v", err))
		return false
	}
	
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		writeLog(fmt.Sprintf("ERROR: Could not start merge: %v", err))
		return false
	}
	
	stderr, err := cmd.StderrPipe()
	if err != nil {
		writeLog(fmt.Sprintf("ERROR: Could not start merge: %v", err))
		return false
	}
	
	if err := cmd.Start(); err != nil {
		writeLog(fmt.Sprintf("ERROR: Could not start KeePassXC-CLI: %v", err))
		return false
	}
	
	// Send password
	fmt.Fprintf(stdin, "%s\n", password)
	stdin.Close()
	
	// Read output
	output, _ := io.ReadAll(stdout)
	errors, _ := io.ReadAll(stderr)
	
	if err := cmd.Wait(); err != nil {
		writeLog(fmt.Sprintf("ERROR: Merge failed: %s", string(errors)))
		return false
	}
	
	if len(output) > 0 && config.Settings.Debug {
		writeLog(string(output))
	}
	
	writeLog("Merge completed successfully. Local file updated.")
	return true
}

func testConnection() bool {
	writeLog("=== Connection Test ===")
	errors := []string{}
	
	// Check KeePassXC
	keepassxc := findExecutable(config.Keepass.KeepassXCPath)
	if keepassxc == "" {
		errors = append(errors, "❌ KeePassXC-CLI not found")
	} else {
		writeLog(fmt.Sprintf("✅ KeePassXC-CLI found: %s", keepassxc))
	}
	
	// Check local database
	if _, err := os.Stat(config.Local.LocalPath); err == nil {
		info, _ := os.Stat(config.Local.LocalPath)
		writeLog(fmt.Sprintf("✅ Local database: %s (%d bytes, modified: %s)", 
			config.Local.LocalPath, info.Size(), info.ModTime().Format("2006-01-02 15:04:05")))
	} else {
		errors = append(errors, fmt.Sprintf("⚠️ Local database not found: %s", config.Local.LocalPath))
	}
	
	// Test connection
	protocol := strings.ToLower(config.FTP.Type)
	writeLog(fmt.Sprintf("Testing connection (%s)...", strings.ToUpper(protocol)))
	
	// Simple connection test (implement based on protocol)
	if protocol == "ftp" || protocol == "sftp" {
		lftpPath := findExecutable("lftp")
		if lftpPath == "" {
			errors = append(errors, "❌ lftp not found")
		} else {
			var url string
			if protocol == "sftp" {
				url = fmt.Sprintf("sftp://%s", config.FTP.Host)
			} else {
				url = fmt.Sprintf("ftp://%s", config.FTP.Host)
			}
			cmd := exec.Command(lftpPath, "-u", fmt.Sprintf("%s,%s", config.FTP.User, config.FTP.Password), url, "-e", "quit")
			if err := cmd.Run(); err != nil {
				errors = append(errors, fmt.Sprintf("❌ Connection failed: %v", err))
			} else {
				writeLog(fmt.Sprintf("✅ Connection to server successful: %s", config.FTP.Host))
			}
		}
	}
	
	writeLog("=== Test completed ===")
	if len(errors) > 0 {
		writeLog("Errors found:")
		for _, e := range errors {
			writeLog(e)
		}
		return false
	}
	
	writeLog("✅ All tests successful!")
	return true
}

func showStatus() {
	writeLog("=== KeePass Sync Status ===")
	
	// Local DB
	if info, err := os.Stat(config.Local.LocalPath); err == nil {
		age := time.Since(info.ModTime())
		writeLog(fmt.Sprintf("Local DB: %s", config.Local.LocalPath))
		writeLog(fmt.Sprintf("  Size: %d bytes (%.2f KB)", info.Size(), float64(info.Size())/1024))
		writeLog(fmt.Sprintf("  Modified: %s (age: %dd %dh)", 
			info.ModTime().Format("2006-01-02 15:04:05"), int(age.Hours()/24), int(age.Hours())%24))
	} else {
		writeLog(fmt.Sprintf("⚠️ Local DB not found: %s", config.Local.LocalPath))
	}
	
	// Backups
	if files, err := filepath.Glob(filepath.Join(config.Local.BackupDir, "*.kdbx")); err == nil {
		writeLog(fmt.Sprintf("Backups: %d found", len(files)))
		// Show top 3
		for i, f := range files {
			if i >= 3 {
				break
			}
			if info, err := os.Stat(f); err == nil {
				writeLog(fmt.Sprintf("  %d. %s (%.2f KB, %s)", i+1, filepath.Base(f), 
					float64(info.Size())/1024, info.ModTime().Format("2006-01-02 15:04:05")))
			}
		}
	}
	
	// Configuration
	writeLog(fmt.Sprintf("Protocol: %s", strings.ToUpper(config.FTP.Type)))
	writeLog(fmt.Sprintf("Server: %s", config.FTP.Host))
	writeLog(fmt.Sprintf("User: %s", config.FTP.User))
	
	// KeePassXC
	if keepassxc := findExecutable(config.Keepass.KeepassXCPath); keepassxc != "" {
		writeLog(fmt.Sprintf("KeePassXC-CLI: ✅ %s", keepassxc))
	} else {
		writeLog("KeePassXC-CLI: ❌ Not found")
	}
}

func performSync() bool {
	keepassxc := findExecutable(config.Keepass.KeepassXCPath)
	if keepassxc == "" {
		writeLog("ERROR: KeePassXC-CLI not found")
		return false
	}
	
	// Backup
	createBackup(config.Local.LocalPath, config.Local.BackupDir)
	cleanupBackups(config.Local.BackupDir, config.Local.MaxBackups)
	
	protocol := strings.ToLower(config.FTP.Type)
	port := config.FTP.Port
	if port == 0 {
		if protocol == "ftp" {
			port = 21
		} else if protocol == "sftp" || protocol == "scp" {
			port = 22
		}
	}
	
	// Download
	if !downloadFile(config.FTP.Host, config.FTP.User, config.FTP.Password, config.FTP.RemotePath,
		config.Local.TempPath, protocol, config.Settings.MaxRetries, config.Settings.RetryDelay,
		port, config.FTP.Share, config.FTP.Domain) {
		return false
	}
	
	// Merge
	if !mergeDatabases(keepassxc, config.Local.LocalPath, config.Local.TempPath, config.Keepass.DatabasePassword) {
		return false
	}
	
	// Upload
	if !uploadFile(config.FTP.Host, config.FTP.User, config.FTP.Password, config.FTP.RemotePath,
		config.Local.LocalPath, protocol, config.Settings.MaxRetries, config.Settings.RetryDelay,
		port, config.FTP.Share, config.FTP.Domain) {
		return false
	}
	
	// Cleanup
	if err := os.Remove(config.Local.TempPath); err == nil {
		writeLog("Temporary file deleted")
	}
	
	writeLog("Synchronization completed.")
	return true
}

func main() {
	// Change to script directory
	if exe, err := os.Executable(); err == nil {
		if dir := filepath.Dir(exe); dir != "" {
			os.Chdir(dir)
			// Go one level up if we're in go/ subdirectory
			if filepath.Base(dir) == "go" {
				os.Chdir("..")
			}
		}
	}
	
	// CLI flags
	testFlag := flag.Bool("test", false, "Test connection without sync")
	statusFlag := flag.Bool("status", false, "Show status")
	watchFlag := flag.Bool("watch", false, "Watch file for changes")
	configFlag := flag.String("config", "config.json", "Alternative config file")
	verboseFlag := flag.Bool("verbose", false, "Verbose output")
	quietFlag := flag.Bool("quiet", false, "Quiet mode (errors only)")
	versionFlag := flag.Bool("version", false, "Show version")
	flag.Parse()
	
	if *versionFlag {
		fmt.Println("KeePass Sync 1.1.0 (Go)")
		return
	}
	
	// Load config
	configFile = *configFlag
	if err := loadConfig(); err != nil {
		fmt.Fprintf(os.Stderr, "ERROR: %v\n", err)
		os.Exit(1)
	}
	
	if *verboseFlag {
		config.Settings.Debug = true
	}
	
	if !*quietFlag {
		writeLog(fmt.Sprintf("=== KeePass Sync - Go ==="))
	}
	
	// Actions
	if *testFlag {
		success := testConnection()
		if success {
			os.Exit(0)
		}
		os.Exit(1)
	}
	
	if *statusFlag {
		showStatus()
		os.Exit(0)
	}
	
	if *watchFlag {
		writeLog("File watching not yet implemented in Go version")
		writeLog("Use Python version for file watching feature")
		os.Exit(1)
	}
	
	// Normal sync
	success := performSync()
	if success {
		os.Exit(0)
	}
	os.Exit(1)
}

